README.md
# 3_Middleman

Middleman](http://middlemanapp.com/) is a static site generator using all the shortcuts and tools in modern web development. It is a ruby (sinatra) application.

Usage
=====

http://middlemanapp.com

Requirements
============
Supports only Ubuntu and debian derivatives at this time
It is expected that ChefDK, chef-client and inspec are installed on the testing platform

The components that this cookbook were tested against were as follows

Chef Development Kit Version: 2.5.3
chef-client version: 13.8.5
berks version: 6.3.1
kitchen version: 1.20.0
inspec version: 1.51.21

The Chef Development Kit may be downloaded at https://downloads.chef.io/chefdk/3.0.36
Inspec may be downloaded at https://www.inspec.io/downloads/

Please note that git is a requirement for the deployment of Chef.

Optionally Vagrant may be downloaded at https://www.vagrantup.com/downloads.html.  
This is not a requirement for the operation of the cookbook however this utility can be
useful if other configuration options are to be considered for the deployment.

Please note that testing was conducted on a Mac utilizing Oracle VirtualBox Manager version 5.2.12.  Downloads for this utility can be found at https://www.virtualbox.org/wiki/Downloads.
Also note that this is an optional utility as testing can be conducted utilizing the chef-client.  Please consult the chef documentation at https://docs.chef.io/ for further details

Testing
=======

This cookbook was developed utilizing test kitchen and Vagrant.  During each run,
the resulting server is deployed to VirtualBox.

To run the test, please be in the 1_MongoDB directory

A `.kitchen.yml` file is provided. Run `kitchen test` to verify this cookbook.

The kitchen test cycle is composed of the following steps:  

o `kitchen converge`.  This creates an instance of the deployed image

o `kitchen verify`.    This command verifies that the deployed instance is operating as expected.

o `kitchen destroy`.   Will remove the deployed instance


Optional commands:

o `kitchen login`      Will allow the user to connect to a deployed image


Environment validation
======================

o Login into the deployed instance utilizing `kitchen login`
o Change directory /home/apache/middleman-blog
o Run 'sudo middleman server'

From your local browser enter the following

http://0.0.0.0:4567 to produce browser output


CHANGELOG.md
# 3_Middleman Cookbook Changelog

# 0.6.1

- Corrected a deploy issue with Middleman

# 0.6.0

- Fixed connection issue with test kitchen

# 0.5.0

- Install gems

  Install thin server and service

# 0.4.0

- Install git

  Install git repository middleman-blog

# 0.3.0

- Install and configure apache version 2

# 0.2.0

- Install ruby

  Install gem

# 0.1.0

- Initial development

  Install dependent packages


spec/unit/recipes/default_spec.rb
#
# Cookbook:: 3_Middleman
# Spec:: default
#
# Copyright:: 2018, The Authors, All Rights Reserved.

require 'spec_helper'

describe '3_Middleman::default' do
  context 'When all attributes are default, on Ubuntu 16.04' do
    let(:chef_run) do
      # for a complete list of available platforms and versions see:
      # https://github.com/customink/fauxhai/blob/master/PLATFORMS.md
      runner = ChefSpec::ServerRunner.new(platform: 'ubuntu', version: '16.04')
      runner.converge(described_recipe)
    end

    it 'converges successfully' do
      expect { chef_run }.to_not raise_error
    end
  end

  context 'When all attributes are default, on CentOS 7.4.1708' do
    let(:chef_run) do
      # for a complete list of available platforms and versions see:
      # https://github.com/customink/fauxhai/blob/master/PLATFORMS.md
      runner = ChefSpec::ServerRunner.new(platform: 'centos', version: '7.4.1708')
      runner.converge(described_recipe)
    end

    it 'converges successfully' do
      expect { chef_run }.to_not raise_error
    end
  end
end


test/integration/default/default_test.rb
# Validate that ruby has been created
describe file('/usr/local/bin/ruby') do
  it { should exist }
end

# Validate that gem has been created
describe file('/usr/local/bin/gem') do
  it { should exist }
end

# Validate that rudy working directory has been created
describe file('/home/vagrant/ruby') do
  it { should_not exist }
end

# Validate that ruby has been installed
describe file('/usr/bin/ruby') do
  it { should exist }
  its('mode') { should cmp '0755' }
end

# Validate that gem has been installed
describe file('/usr/bin/gem') do
  it { should exist }
  its('mode') { should cmp '0755' }
end

# Validate that apache2 has been installed
describe file('/etc/apache2/apache2.conf') do
  it { should exist }
end

# Validate that apache2 configuration file proxy_load has been installed
describe file('/etc/apache2/mods-enabled/proxy_http.load') do
  it { should exist }
end

# Validate that apache2 configuration file rewrite has been installed
describe file('/etc/apache2/mods-enabled/rewrite.load') do
  it { should exist }
end

# Validate that apache2 configuration file rewrite has been installed
describe file('/etc/apache2/sites-enabled/blog.conf') do
  it { should exist }
end

# Validate that apache2 configuration file rewrite has been installed
describe file('/etc/apache2/sites-enabled/000-default.conf') do
  it { should_not exist }
end

# Validate that git has been installed
describe file('/usr/bin/git') do
  it { should exist }
  its('mode') { should cmp '0755' }
end

# Validate that middleman blog exists
describe file('/home/apache/middleman-blog') do
  it { should exist }
end

# Validate that gem bundler has been installed
describe file('/usr/local/lib/ruby/gems/2.1.0/gems/bundler-1.16.1') do
  it { should exist }
  its('mode') { should cmp '0755' }
end

# Required working directory
describe file('/home/apache/.bundle/cache/compact_index') do
  it { should exist }
end

# Middleman has been deployed
describe file('/var/www/middleman/config.rb') do
  it { should exist }
end


metadata.rb
name '3_Middleman'
maintainer 'Evan Haxton'
maintainer_email 'ehaxton@gmail.com'
license 'apachev2'
description 'Installs/Configures Middleman ruby application'
long_description 'Please refer to full description in README.md'
version '0.6.1'
chef_version '>= 12.14' if respond_to?(:chef_version)

# The `issues_url` points to the location where issues for this cookbook are
# tracked.  A `View Issues` link will be displayed on this cookbook's page when
# uploaded to a Supermarket.
#
# issues_url 'https://github.com/<insert_org_here>/3_Middleman/issues'

# The `source_url` points to the development repository for this cookbook.  A
# `View Source` link will be displayed on this cookbook's page when uploaded to
# a Supermarket.
#
# source_url 'https://github.com/<insert_org_here>/3_Middleman'

depends 'apt', '~> 7.0.0'
depends 'tar', '~> 2.1.1'
depends 'sudo', '~> 5.4.0'


recipes/default.rb
#
# Cookbook:: 3_Middleman
# Recipe:: default
#
# Copyright:: 2018, Evan Haxton, All Rights Reserved.

# apt-get update
apt_update 'update'

# apt-get install build-essential libssl-dev libyaml-dev libreadline-dev openssl curl git-core zlib1g-dev bison libxml2-dev libxslt1-dev libcurl4-openssl-dev nodejs libsqlite3-dev sqlite3
apt_package %w( build-essential libssl-dev libyaml-dev libreadline-dev openssl curl git-core zlib1g-dev bison libxml2-dev libxslt1-dev libcurl4-openssl-dev nodejs libsqlite3-dev sqlite3) do
  action :install
end

# make working directory
directory 'ruby' do
  mode '0755'
  path '/home/vagrant/ruby'
  action :create
end

# download tarball
# wget http://cache.ruby-lang.org/pub/ruby/2.1/ruby-2.1.3.tar.gz
remote_file '/home/vagrant/ruby/ruby-2.1.3.tar.gz' do
  source 'http://cache.ruby-lang.org/pub/ruby/2.1/ruby-2.1.3.tar.gz'
  not_if { ::File.exist?('/usr/local/bin/ruby') }
end

# extract tarball to /home/vagrant/ruby
tar_extract '/home/vagrant/ruby/ruby-2.1.3.tar.gz' do
  action :extract_local
  target_dir '/home/vagrant/ruby'
  notifies :run, 'execute[build_ruby]', :immediately
  creates '/usr/bin/ruby'
  not_if { ::File.exist?('/usr/local/bin/ruby') }
end

# Build ruby
# cd ruby-2.1.3 & ./configure & make install
execute 'build_ruby' do
  command 'sudo ./configure && sudo make install'
  cwd '/home/vagrant/ruby/ruby-2.1.3'
  not_if { ::File.exist?('/usr/local/bin/ruby') }
end

# rm -rf ~/ruby
directory '/home/vagrant/ruby' do
  recursive true
  action :delete
  only_if { ::File.exist?('/home/vagrant/ruby') }
end

# cp /usr/local/bin/ruby /usr/bin/ruby
file '/usr/bin/ruby' do
  owner 'root'
  group 'root'
  mode 0755
  lazy { content ::File.open('/usr/local/bin/ruby').read }
  action :create
  only_if { ::File.exist?('/usr/local/bin/ruby') }
end

# cp /usr/local/bin/gem /usr/bin/gem
file '/usr/bin/gem' do
  owner 'root'
  group 'root'
  mode 0755
  lazy { content ::File.open('/usr/local/bin/gem').read }
  action :create
  only_if { ::File.exist?('/usr/local/bin/gem') }
end

# apt-get install apache2
apt_package 'apache2'

# Configure apache2

# a2enmod proxy_http
execute 'proxy_http' do
  command 'sudo a2enmod proxy_http'
  not_if { ::File.exist?('/etc/apache2/mods-enabled/proxy_http.load') }
end

# a2enmod rewrite
execute 'proxy_http' do
  command 'sudo a2enmod rewrite'
  not_if { ::File.exist?('/etc/apache2/mods-enabled/rewrite.load') }
end

# cp blog.conf /etc/apache2/sites-enabled/blog.conf
template '/etc/apache2/sites-enabled/blog.conf' do
  source 'blog.conf.erb'
end

# rm /etc/apache2/sites-enabled/000-default.conf
file '/etc/apache2/sites-enabled/000-default.conf' do
  action :delete
  manage_symlink_source true
  only_if { ::File.exist?('/etc/apache2/sites-enabled/000-default.conf') }
end

# Restart apache
service 'apache2' do
  action :reload
end

# Install Git
# apt-get install git
apt_package 'git'

# Create apache user
user 'apache' do
  manage_home true
  comment 'Apache User'
  home '/home/apache'
  shell '/bin/bash'
  password 'apache'
end

sudo 'apache' do
  user 'apache'
  nopasswd true
end

# Install project dependencies

# Make /home/apache writable
directory '/home/apache' do
  mode '0757'
  action :create
end

# make working directory for bundle install
directory '/home/apache/.bundle/cache/compact_index' do
  mode '1777'
  recursive true
  owner 'apache'
  action :create
  not_if { ::File.exist?('/home/apache/.bundle/cache/compact_index') }
end

# make working directory for bundle install
directory '/tmp' do
  mode '1777'
  recursive true
  action :create
end

# Install Middleman blog repository
# git clone https://github.com/learnchef/middleman-blog.git
git '/home/apache/middleman-blog' do
  repository 'https://github.com/learnchef/middleman-blog.git'
  revision 'master'
  user 'apache'
  action :checkout
  not_if { ::File.exist?('/home/apache/middleman-blog') }
end

# Install Bundler
# gem install bundler
gem_package 'bundler' do
  version '1.16.1'
  action :install
  not_if { ::File.exist?('/usr/local/lib/ruby/gems/2.1.0/gems/bundler-1.16.1') }
end

# Create apache user
user 'apache' do
  manage_home true
  comment 'Apache User'
  home '/home/apache'
  shell '/bin/bash'
  password 'apache'
end

sudo 'apache' do
  user 'apache'
  nopasswd true
end

# bundle install
execute 'bundle_install' do
  command 'bundle install'
  environment 'tmpdir' => '/tmp'
  cwd '/home/apache/middleman-blog'
  user 'apache'
  not_if { ::File.exist?('/usr/local/lib/ruby/gems/2.1.0/gems/middleman-3.1.0') }
end

# establish configuration file
template '/etc/blog.yml' do
  source 'blog.yml.erb'
  variables project_install_directory: '/var/www/middleman'
  not_if { ::File.exist?('/etc/blog.yml') }
end

# make working directory
directory 'middleman' do
  mode '0755'
  path '/var/www/middleman'
  action :create
end

template '/etc/init.d/thin' do
  source 'thin.erb'
  mode '0755'
  variables home_directory: '/var/www/middleman'
  not_if { ::File.exist?('/etc/init.d/thin') }
end

# thin install /usr/sbin/update-rc.d -f thin defaults
execute 'thin install' do
  command '/usr/sbin/update-rc.d -f thin defaults'
end

# restart service
service 'thin' do
  action :restart
end

# deploy middleman
execute 'middleman deploy' do
  command 'sudo middleman init /var/www/middleman'
  not_if { ::File.exist?('/var/www/middleman/config.rb') }
end
